controller/engine_controller.go

```diff
@@ -1394,5 +1394,12 @@ func (ec *EngineController) UpgradeEngineProcess(e *longhorn.Engine) error {
 }
 
 func (ec *EngineController) isResponsibleFor(e *longhorn.Engine) bool {
-	return isControllerResponsibleFor(ec.controllerID, ec.ds, e.Name, e.Spec.NodeID, e.Status.OwnerID)
+	preferredOwnerID := ""
+	v, err := ec.ds.GetVolume(e.Spec.VolumeName)
+	if err != nil {
+		preferredOwnerID = e.Spec.NodeID
+	} else {
+		preferredOwnerID = v.Status.OwnerID
+	}
```

Comment:
Assume that:
* volume is attaching to `node-1`. 
* `v.Status.OwnerID` = `node-1`, 
* `v.Spec.NodeID` = `node-1`, 
* `e.Spec.NodID` = `node-1`, 
* `e.Status. OwnerID ` = `node-1`

Now if the engine image is suddenly unavailable on `node-1` (maybe the engine image daemonset is evicted for some reasons), the `v.Status.OwnerID` changes to a node that has engine image binary, `node-2`, so that it can do backup and snapshot operations. The `e.Spec. OwnerID `  must also change to a node that has engine image so we can refresh the engine monitor and update the engine status. Since we already have the logic to switch  `v.Status.OwnerID`, we can just use `v.Status.OwnerID` as the `preferredOwnerID`. I think it also cleaner to keep engine `v.Status.OwnerID` and `e.Spec. OwnerID ` on the same node

Reply:
Ideally, I don't want to use `v.Status.OwnerID` outside of the volume controller, which is designed for determining if a volume can be handled by one of its controller or not. But this logic can make the engine and the volume strongly consistent. Then I am fine with either. BTW, can this modification handle the live migration feature, which will have 2 running engines for 1 running volume?

Reply:
> BTW, can this modification handle the live migration feature, which will have 2 running engines for 1 running volume?

I am not sure about the `live migration feature`. cc @joshimoo

Reply:
@joshimoo Can you take a brief look to confirm that this won't break the live migration feature?

Reply:
I cannot see the above lines, I will have another look once the code for ownership transfer is moved into the isResponsbile check as mentioned in my pr comments.